core/iwasm/common/wasm_shared_memory.c

```diff
@@ -87,6 +87,42 @@ search_module(WASMModuleCommon *module)
     return NULL;
 }
 
+static void
+notify_stale_threads(void *key, void *value, void *user_data)
+{
+    WASMModuleInstanceCommon *module_inst =
+        (WASMModuleInstanceCommon *)user_data;
+    wasm_runtime_atomic_notify(module_inst, key, UINT32_MAX);
+    bh_hash_map_remove(wait_map, key, NULL, NULL);
+}
+
+static void
+create_copy_of_wait_map(void *key, void *value, void *map_copy)
+{
+    bh_hash_map_insert((HashMap *)map_copy, key, value);
+}
+
+void
+notify_stale_threads_on_exception(WASMModuleInstanceCommon *module_inst)
+{
+
+    static HashMap *wait_map_without_lock;
+    wait_map_without_lock = bh_hash_map_create(
+        32, false, (HashFunc)wait_address_hash,
+        (KeyEqualFunc)wait_address_equal, NULL, destroy_wait_info);
+    /**
+     * create a copy of wait_map with no lock
+     * this allows invoking methods in the callback
+     * which require acquiring lock on the original map
+     */
+    bh_hash_map_traverse(wait_map, create_copy_of_wait_map,
+                         wait_map_without_lock);
+
+    // terminate stale threads
+    bh_hash_map_traverse(wait_map_without_lock, notify_stale_threads,
+                         module_inst);
+}
```

Comment:
we should probably call `bh_hash_map_destroy` on `wait_map_without_lock` at the end